### ISSUE TITLE
linearsolve using the transpose of the factorization

### ISSUE THREAD

**Issue by MKAbdElrahman** [SciML/LinearSolve.jl#92]
For defining rrules,  I need to make use of the factorization in the forward pass.  The rrules is also a linearsolve with the transpose of the linear system, how to avoid  defining a new LinearProblem ? 

This my current code, the LinearSolve interface with make it make use other algorithms  and  more clean  

```julia

function ChainRulesCore.rrule(::typeof(efield), sim::Simulation,ϵ)
    linsys = ConstructLinearSystem(sim, ϵ )
    A ,b = linsys.A , linsys.b
    x = similar(b); x_adj = similar(b)
    F = lu(A)
    LinearAlgebra.ldiv!(x, F, b)
    sim.E = x
    function efield_pullback(ȳ)
        LinearAlgebra.ldiv!(x_adj, transpose(F), conj.(ȳ))
        f̄ = NoTangent()
        f̄oo =  NoTangent()
        ϵoo =  real(x .* x_adj)
        return f̄, f̄oo , ϵoo
    end
    return x, efield_pullback
end


Thanks!
```

**Comment by ChrisRackauckas** [comment]
oh yes, this would be good to add. Now it might be hard for Krylov methods because it will require that the operator has an adjoint defined, and many times they might be defined Jacobian-free, that will just throw an appropriate error if it's not defined.

I think the right thing to do would be to add a boolean to the LinearProblem `transpose=false` by default, and then we can setup the algorithms to specialize on this. Many, such as Pardiso, have a lazy transpose option in the solver so we'd use that bool to flip the option.

**Comment by vpuri3** [comment]
@MKAbdElrahman did the issue get resolved?

**Comment by ChrisRackauckas** [comment]
It did not

**Comment by vpuri3** [comment]
In `LinearCache`, we can add a flag `symmetric` defaulting to `false` and a field `Atransp` defaulting to `Adjoint(A)`, the lazy wrapper. And then to solve the adjoint problem via

`solve(prob, alg, adjoint=true)`

**Comment by vpuri3** [comment]
maybe change terminology to make it more clear for complex eltypes
- `symmetric` --> `selfadjoint`
- `Atransp` --> `Aadjoint`

**Comment by vpuri3** [comment]
@ChrisRackauckas is `Adjoint(::DiffEqArrayOperator)` defined?

**Comment by ChrisRackauckas** [comment]
No. What needs to be done is the operator interface documentation should get a note about `adjoint(::SciMLOperator)` as being a part of the (optional) interface, required for reverse mode automatic differentiation. Then, `adjoint(::DiffEqArrayOperator)` should be added to SciMLBase by just taking the adjoint of the internal array. Many other operators would have to be handled though, but that can be done over time.

**Comment by vpuri3** [comment]
sounds good, is `solve(prob, alg, adjoint=true)`the standard interface for solving a joint problem in diffeq?

**Comment by ChrisRackauckas** [comment]
Not in solve, it's not a solve level thing. It changes the result, so it's not a solver control but instead something to do with the problem. The real question is whether the answer is for it to just be `LinearProblem(A',b)` or `LinearProblem(A,b,adjoint=true)`. The advantage of the latter is that it makes it easier to do general implementations of adjoints for operators which only define `A*x`, since then it can use AbstractDifferentiation.jl behind the scenes. However, you don't have to do that if `adjoint(A)` already exists. So we need `has_adjoint(A::AbstractSciMLOperator)` and `adjoint(A::AbstractSciMLOperator)` in the interface in order to write it effectively. Also, I'm not so sure we want the AD overloads though, just because of the dependencies that would give. So it would need to do something like DiffEqSensitivity, where by default an error is raised saying to `using DiffEqSensitivity` to get the adjoint overloads, unless `has_adjoint(A)`. This complexity is eliminated if you just assume `LinearProblem(A',b)` is the way to do it.

**Comment by vpuri3** [comment]
`LinearProblem(prob, alg, adjoint=true)` would be good for factorizations since adjoint of `LinearAlgebra.Factorization` is defined. For iterative methods, etc we can fallback to adjoint wrapper with a `has_adjoint` check

**Comment by briochemc** [comment]
I could not find anything in the docs but is there an `adjoint` switch now to avoid redefining a new `LinearProblem`? Or is this still an open issue? I have a "simpler" use case where I define a problem, solve it, and then need to solve its adjoint too, typically like:
```julia
prob = init(LinearProblem(A, b))
sol = solve!(prob).u
adjointsol

**Comment by ChrisRackauckas** [comment]
There is not. For `Matrix` it's handled just by `adjoint`, so it's more a SciMLOperators issue.

**Comment by AJ0070** [comment]
The reuse machinery for this already exists in LinearSolve, but only reverse-mode AD can reach it. `_adjoint_factorization_solve` handles about 22 factorization types across four reuse strategies, and it is called only from the ChainRulesCore, Enzyme and Mooncake extensions.

For a user writing an `rrule` by hand, as in the original post, there is no public way in, so the adjoint solve refactorizes. On a 400x400 dense LU:

```
separate LinearProblem(A', c)   0.425 s   (refactorizes)
reusing the factorization       0.164 s   (2.6x faster, same answer)
```

Is exposing that reuse something you would take, or do you still see this as belonging on the SciMLOperators side? If the former, the shape matters: you ruled out a `solve`-level flag earlier, so it would presumably be a cache-level entry point rather than a keyword.

**Comment by AJ0070** [comment]
@ChrisRackauckas gentle ping on the question above, since a decision here is what blocks any work on this one.

The reuse already exists and is exercised by the AD rules, it just has no public entry point, and it measured 2.6x on a 400x400 dense LU. You ruled out a `solve` level flag earlier in this thread, so the open question is whether you want it exposed at all, and if so whether as a cache level entry point rather than the `LinearProblem(A, b, adjoint = true)` shape discussed above.

**Comment by ChrisRackauckas** [comment]
I have come around to see that it should be a solve level thing, since you want to do this on an already created factorization object. `adjointsol = solve!(prob; adjoint=true).u` , and the AD overloads should use this route instead of what they hardcode.
